### DISCUSSION TITLE
MI wireless switch (WXKG01LM) ZigBee Data (Payload)

### DISCUSSION THREAD

**Discussion by rmeissne** [Koenkk/zigbee2mqtt#13470] [Devices]
Hello,

I have captures messages received from the ZigBee adapter when pressing the button of the switch:

fe1b4481 0000 0600 0e5f 01 01 00 78 0024c97a0000 07 **18f0 0a 000010000e** 5f1ddf (1x press when button is down)
fe1b4481 0000 0600 0e5f 01 01 00 6e 00a9cb7a0000 07 **18f1 0a 000010010e** 5f1d46 (1x press when button got released)

fe1b4481 0000 0600 0e5f 01 01 00 6e 0068987b0000 07 **18f2 0a 008020020e** 5f1d65 (2x press)

fe1b4481 0000 0600 0e5f 01 01 00 63 00f5007c0000 07 **18f3 0a 008020030e** 5f1d6a (3x press)

fe1b4481 0000 0600 0e5f 01 01 00 69 00586e7c0000 07 **18f4 0a 008020040e** 5f1da3 (4x press)

fe1b4481 0000 0600 0e5f 01 01 00 6e 00e9e97c0000 07 **18f5 0a 008020800e** 5f1d17 (5x press)

fe1b4481 0000 0600 0e5f 01 01 00 63 0047517d0000 07 **18f6 0a 008020800e** 5f1d0e (6x press)

All are with cluster id 0006. We can see the link quality value. Source Address (of the button) which is 5f0e. The payload of the actual button data is marked in bold.

How are you guys doing this? Are just guessing the values? Certain values are clear: 18f0, 18f1 etc. this is the a message count. No idea why one would need this value. 0/1 you see when pressed and when released. 02, 03, 04 the repetition. After the count there is always a 0a. For the single messages 000010 seems to be fixed with an additional 00 for being pressed or a 01 for being released. Other messages have 008020 followed by 02/03/04. If you press 5x or 6x you will get 80 as the next value.

Does anybody know where this is documented? What is 0e? It is not fixed, as it does change. Where is the battery state/voltage. Guess there is another message for it as the plan was to make this even more complicated.

Thanks.

Regards,
Robert

**Comment by rmeissne** [comment]
Just captured another message:

fe314481 0000 0000   0e5f 01 01 00 6b 00 b5343c00 00 1d **1c3412000a02ff4c0600100121c40b21a81324090000000021900620500e** 5f 1d 2a

This is cluster 0000 (basic). Payload is again in bold. We can see c40b (0bc4) which equals 3012 - which is apparently the voltage level (as per my Zigbee2Mqtt log). Looks like this message does hold the voltage level.

You guys are somehow parsing/converting the data. Where can I find this information please.

Thanks,
Robert

**Reply by Koenkk** [reply]
See https://github.com/Koenkk/zigbee-herdsman/blob/62f6ff39d921cc12bbaccb7ff6ee6cac8c0b438b/src/zcl/buffaloZcl.ts#L189

**Reply by rmeissne** [reply]
Thanks a lot. OK, I did have a look - but to be honest - I still don't know what the fields are :) Is this somewhere documented? Somebody was clever enough to figure this out otherwise nobody could parse this information.

**Reply by Koenkk** [reply]
No this is not documented somewhere

**Reply by rmeissne** [reply]
OK, but how do you now figure out the voltage level form a device? You need to put "sense" into the messages received from the devices, right? Not sure, maybe I am confused here.

**Reply by Koenkk** [reply]
It is in the message indeed, after the parsing by buffaloZcl it is attribute 1 of the struct: https://github.com/Koenkk/zigbee-herdsman-converters/blob/105b2f20b1879ca2b824023657d8fae303520f55/lib/xiaomi.js#L149
